### PR TITLE
PWGHF: Developements for Dstar V0 resonances analysis workflow

### PR DESCRIPTION
**Pull request by Luca610**
This PR implements a few changes necessary for the Ds resonances analysis in the Dstar V0 decay channel.

- utilsAnalysis.h: added a function for pT differential parametrized cut for 2 prong candidates to reproduce the trigger selection 
cuts and improved form of previous PR for 3 prong candidates.
- candidateSelectorDstar: implemented the trigger-like mass cut on the D0 mass
- candidateSelectorDplus: adapted cut function to new cleaner version
- ReducedDataModel.h: modified reduced collision table to include number of primary contributors, event selection bitmask and Bz for future developements
- dataCreatorCharmReso and dataCreatorCharmHad: modified to fill new collision table

### CODE REVIEW

**Comment by vkucera** [comment]
Hi @Luca610 , I see that you marked my comments as resolved. Please don't do that. The reasons are explained in the contribution guidelines. You also did not make any changes in the PR. Can you please comment?

File: PWGHF/Utils/utilsAnalysis.h [hunk 1]
```diff
@@ -111,32 +109,46 @@ bool isSelectedTrackTpcQuality(T const& track, const int tpcNClustersFoundMin, c
...
+  o2::framework::Configurable<float> ptMassCut3ProngMax{"ptMassCut3ProngMax", 8., "Max pt to apply pt differential cut for 3 prong candidates"};
+  o2::framework::Configurable<float> ptMassCut2ProngMax{"ptMassCut2ProngMax", 8., "Max pt to apply pt differential cut for 2 prong candidates"};
   o2::framework::Configurable<o2::framework::LabeledArray<float>> deltaMassPars3Prong{"deltaMassPars3Prong", {defaultDeltaMassPars3Prong[0], 2, {"constant", "linear"}}, "delta mass parameters for HF 3 prong trigger mass cut"};
+  o2::framework::Configurable<o2::framework::LabeledArray<float>> deltaMassPars2Prong{"deltaMassPars2Prong", {defaultDeltaMassPars2Prong[0], 2, {"constant", "linear"}}, "delta mass parameters for HF 2 prong trigger mass cut"};
   o2::framework::Configurable<o2::framework::LabeledArray<float>> sigmaPars3Prong{"sigmaPars3Prong", {defaultSigmaPars3Prong[0], 2, {"constant", "linear"}}, "sigma parameters for HF 3 prong trigger mass cut"};
+  o2::framework::Configurable<o2::framework::LabeledArray<float>> sigmaPars2Prong{"sigmaPars2Prong", {defaultSigmaPars2Prong[0], 2, {"constant", "linear"}}, "sigma parameters for HF 2 prong trigger mass cut"};
```
Comment (vkucera) [inline PWGHF/Utils/utilsAnalysis.h:129]:
```suggestion
  o2::framework::Configurable<float> nSigma3ProngMax{"nSigma3ProngMax", 2, "Maximum number of sigmas for pT-differential mass cut for 3-prong candidates"};
  o2::framework::Configurable<float> nSigma2ProngMax{"nSigma2ProngMax", 2, "Maximum number of sigmas for pT-differential mass cut for 2-prong candidates"};
  o2::framework::Configurable<float> ptDeltaMass3ProngMax{"ptDeltaMass3ProngMax", 10., "Max pT to apply delta mass shift to PDG mass value for 3-prong candidates"};
  o2::framework::Configurable<float> ptDeltaMass2ProngMax{"ptDeltaMass2ProngMax", 10., "Max pT to apply delta mass shift to PDG mass value for 2-prong candidates"};
  o2::framework::Configurable<float> ptMassCut3ProngMax{"ptMassCut3ProngMax", 8., "Max pT to apply pT-differential cut for 3-prong candidates"};
  o2::framework::Configurable<float> ptMassCut2ProngMax{"ptMassCut2ProngMax", 8., "Max pT to apply pT-differential cut for 2-prong candidates"};
  o2::framework::Configurable<o2::framework::LabeledArray<float>> deltaMassPars3Prong{"deltaMassPars3Prong", {defaultDeltaMassPars3Prong[0], 2, {"constant", "linear"}}, "delta mass parameters for HF 3-prong trigger mass cut"};
  o2::framework::Configurable<o2::framework::LabeledArray<float>> deltaMassPars2Prong{"deltaMassPars2Prong", {defaultDeltaMassPars2Prong[0], 2, {"constant", "linear"}}, "delta mass parameters for HF 2-prong trigger mass cut"};
  o2::framework::Configurable<o2::framework::LabeledArray<float>> sigmaPars3Prong{"sigmaPars3Prong", {defaultSigmaPars3Prong[0], 2, {"constant", "linear"}}, "sigma parameters for HF 3-prong trigger mass cut"};
  o2::framework::Configurable<o2::framework::LabeledArray<float>> sigmaPars2Prong{"sigmaPars2Prong", {defaultSigmaPars2Prong[0], 2, {"constant", "linear"}}, "sigma parameters for HF 2-prong trigger mass cut"};
```

File: PWGHF/Utils/utilsAnalysis.h [hunk 2]
```diff
@@ -111,32 +109,46 @@ bool isSelectedTrackTpcQuality(T const& track, const int tpcNClustersFoundMin, c
...
-  /// \param pTrackSameChargeFirst is the opposite charge track momentum
-  /// \param ptD is the pt of the D+ meson candidate
-  /// \return true if D+ passes selection
-  bool isSelectedDplusInMassRange(const float& invMassDplus, const float& ptD)
+  /// Mass selection of 2 or 3 prong canidates in triggered data analysis
+  /// \param invMass is the invariant mass of the candidate
+  /// \param pdgMass is the pdg Mass of the candidate particle
+  /// \param pt is the pt of the candidate
+  /// \return true if candidate passes selection
+  template <bool is3Prong>
```
Comment (vkucera) [inline PWGHF/Utils/utilsAnalysis.h:136]:
```suggestion
  /// Mass selection of 2-prong or 3-prong candidates in triggered data analysis
  /// \tparam is3Prong switch between 2-prong and 3-prong selection
  /// \param invMass invariant mass of the candidate
  /// \param pdgMass PDG mass of the candidate particle
  /// \param pt pT of the candidate
  /// \return true if candidate passes selection
  template <bool is3Prong>
```

File: PWGHF/Utils/utilsAnalysis.h [hunk 3]
```diff
@@ -111,32 +109,46 @@ bool isSelectedTrackTpcQuality(T const& track, const int tpcNClustersFoundMin, c
...
-
-    if (std::fabs(invMassDplus - peakMean) > nSigma * peakWidth && ptD < pTMaxMassCut) {
-      return false;
+    float peakMean{0.};
+    float peakWidth{0.};
+    float ptMassCutMax{0.};
+    if constexpr (is3Prong) {
+      peakMean = (pt < ptDeltaMass3ProngMax) ? ((pdgMass + deltaMassPars3Prong->get("constant")) + deltaMassPars3Prong->get("linear") * pt) : pdgMass;
+      peakWidth = sigmaPars3Prong->get("constant") + sigmaPars3Prong->get("linear") * pt;
+      ptMassCutMax = ptMassCut3ProngMax;
```
Comment (vkucera) [inline PWGHF/Utils/utilsAnalysis.h:145]: You can reduce the branching to its minimum by assigning `nSigmaMax`, `ptDeltaMassMax`, `deltaMassPars`, `sigmaPars` in the way you do it for `ptMassCutMax`. The code itself will then work for both scenarios.

File: PWGHF/Utils/utilsAnalysis.h [hunk 4]
```diff
@@ -111,32 +109,46 @@ bool isSelectedTrackTpcQuality(T const& track, const int tpcNClustersFoundMin, c
...
+    } else {
+      float peakMean = (pt < ptDeltaMass2ProngMax) ? ((pdgMass + deltaMassPars2Prong->get("constant")) + deltaMassPars2Prong->get("linear") * pt) : pdgMass;
+      float peakWidth = sigmaPars2Prong->get("constant") + sigmaPars2Prong->get("linear") * pt;
```
[the remark this answers is not in the document]
Comment (Luca610) [reply PWGHF/Utils/utilsAnalysis.h:148]: Thank you for the catch!

**Comment by Luca610** [comment]
Hi @vkucera, thanks again for your suggestions, I implemented all of them with the latest commit

**Review by vkucera — APPROVED** [review]
Thanks @Luca610 for all the changes.

**Pull request merged**